### PR TITLE
[PIM-5620]Remove deprecated import system

### PR DESCRIPTION
**Pull request by aRn0D**
# Diff :  547 files changed, 12636 insertions(+), 29534 deletions(-)

# TransformBundle:
- [x] `Flat` normalizer/denormalizers is moved to Connector (csv format)
- [x] `Structured/json` normalizers/denormalizers is moved to Catalog (standard format)
- [x] Remove all the content of this bundle
    

# BaseConnectorBundle:

Remove those namespaces:
- [x] `/Pim/Bundle/BaseConnectorBundle/Reader/File`, except `YamlReader` still used by the job `fixtures_job_yml` (InstallerBundle)
- [x] `Pim/Bundle/BaseConnectorBundle/Processor/Denormalization`
- [x] `Pim/Bundle/BaseConnectorBundle/Writer/DirectToDB/MongoDB`
- [x] `Pim/Bundle/BaseConnectorBundle/Writer/Doctrine`
- [x]  Remove few classes in `Pim/Bundle/BaseConnectorBundle/Processor`
still classes to remove : 
- pim_base_connector.processor.csv_serializer.heterogeneous
- pim_base_connector.processor.dummy
- pim_base_connector.processor.tag.class
- [x] Cherry pick Pierre's commits

#  PimEnterpriseBaseConnectorBundle
- [x]  Remove the whole PimEnterpriseBaseConnectorBundle, related PR https://github.com/akeneo/pim-enterprise-dev/pull/1734/files

# Behat
- [x]  Remove deprecated scenarios https://github.com/akeneo/pim-community-dev/tree/master/features/deprecated-behat/import (used by the deprecated builds on CI)

# Technical doc
- [x]  Write BC breaks in the changelog + add all the sed in the upgrade.md
- [x]  Complete the doc http://docs.akeneo.com/latest/reference/import_export/index.html to explain that in 1.6 all imports use the new system
- [x]  Change the cookbook http://docs.akeneo.com/latest/cookbook/setup_data/customize_installer.html 

Doc pr : https://github.com/akeneo/pim-docs/pull/262

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | yes
| Added Behats                      | no
| Changelog updated             | yes
| Review and 2 GTM              | yes
| Migration script                  | -
| Tech Doc                          | yes

### CODE REVIEW

**Comment by nidup** [comment]
would keep subfolder for Normalizer and Denormalizer because if we need to add another format we'll be a bit screwed, would say:
 - Pim\Component\Connector\Denormalizer\Flat
 - Pim\Component\Connector\Normalizer\Flat
 - Pim\Component\Catalog\Normalizer\Structured (would love name this one /Standard because it's more meaningful but we already use Structured in several other places)

**Comment by nidup** [comment]
About Normalizers again we use inheritance for almost nothing "class AttributeNormalizer extends BaseNormalizer", at least Connector extends Catalog which is quite ok. I know it's not the scope of this PR but if we have buffer we can drop all these useless overrides.

About Flat Denormalizer, for reminder, we use this only in EE in ProductReverter to do:
```
        $revertedObject = $this->denormalizer->denormalize(
            $data,
            $class,
            'csv',
            [ 'entity' => $currentObject ]
        );
```

If we'd use the standard format for version storage we could use the updater to revert a version and purely drop this set of denormalizers

**Comment by nidup** [comment]
The class MetricConverter could be moved in Catalog, it provides a business behavior, only issue is it depends on "use Akeneo\Bundle\MeasureBundle\Convert\MeasureConverter" so we'll have a coupling violation. The MeasureBundle should be moved into our main repository before to be splitted into component + bundle, the MeasureConverter does not depends at all on Symfony. When we'll re-work product export for ExportBuilder we'll be face of the same issue, as we'll need to use it in the future new product reader which will be located in Connector component. So I would move the class now and in Catalog component assuming to have a violation due to the use of the MeasureBundle.

Ok for the move of the deprecated DoctrineCache, we'll need to finish the installer re-work before to drop it (not in this PR).

Ok for the move of the deprecated CacheClearer, it's only indirectly used by the RefreshCommand (versioning command).

**Comment by nidup** [comment]
The list of sed commands is not enough to explain the change, at least, I would add a small sentence in the upgrade-1.6.md with a link to the @pierallard's new doc (about to be merged).

I would ideally add a "before / after" or "how to migrate" deprecated imports chapter.

**Comment by nidup** [comment]
:fire: Great PR :fire: the API is far more clearer this way, comments to be discussed then GTM.

**Comment by jjanvier** [comment]
Agree with @nidup about the (de)normalizers of the _Catalog_ component. We should put them in a sub-folder. Personally, I would name this sub-folder, and the normalizers type `pivot` as:
* it represents a full object
* this is what should be passed into the `Updaters`

But maybe we could do this in another PR, as this one is already huge.

About the `MetricConverter`, yes, we could move it even it introduced a new violation, that's no big deal. And when you look at this class, it has nothing to do with connectors. This is pure catalog stuff imvho.

Again, this could be done in another PR.

**Comment by juliensnz** [comment]
Disclaimer: what will follow doesn't mean that it should be done in this PR but more what I think should be done before the 1.6 release.

I think that we should drop the Flat/Structured folders. Normalizer/Denormalizer in the PIM should always work with the pivot/standard format and stick to it. We should also drop the currently supported formats (xml, json and csv) to only support 'standard' (or 'pivot'). 

We now have the weapons to do it: converters, updaters and proper normalizers. Dropping totally the flat normalizers/denormalizers will reduce the code complexity and ease a lot our workflow. We should implement the standard -> flat converters and use them in export and versioning. With that, we could get rid of all this the old flat normalizer/denormalizer.

As those remark cannot be done in this PR, I think that we should agree on an end goal (for the 1.6) and make our decisions based on it for this PR to avoid multiple rework on this subject.

Maybe the best way to handle this question is to talk about it IRL so don't hesitate if you want to !

**Comment by aRn0D** [comment]
Structured normalizers are moved to `Pim\Component\*\Normalizer\Structured` for now it will rework in another PR!

File: features/enrich/variant-group/add_products/add_products_history_updated_with_context.feature [hunk 1]
```diff
@@ -19,6 +69,7 @@ Feature: Add products to a variant group
       | SANDAL | manufacturer | Converse    |        |       |
       | SANDAL | name         | EN name     | en_US  |       |
       | SANDAL | comment      | New comment |        |       |
+    And I logout
     And I am logged in as "Julia"
 
   Scenario: Successfully add products in variant groups, history should be updated with a variant group context
```
Comment (SamirBoulil) [inline features/enrich/variant-group/add_products/add_products_history_updated_with_context.feature]: Shouldn't it be called "Variant group history should be updated successfully" ? It just does not add products or variants here, background does no ?

**Comment by ClementGautier** [comment]
Little things:

* Use addAxisAttribute in GroupUpdater::setAxis instead of setAxis could make the code cleaner
* You have some checkstyle issue (in `src/Pim/Component/Connector/ArrayConverter/Flat/AttributeStandardConverter.php` you basicly changed `(cast) $maVar` for `(cast)$maVar` which should fail on the phpcs. I don't know why travis didn't noticed it. But, as it is coding style and we agreed that coding standard should be checked only be automated tools I'm ok with not fixing it now.

GTM otherwise, good job.

File: spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterSerializerPassSpec.php [hunk 2]
```diff
@@ -109,17 +95,20 @@ function it_sorts_arguments_by_priority(
...
-
-        $definition->setArguments(
-            [
-                [$barNormalizer, $fooNormalizer],
-                [$bazEncoder, $quxEncoder]
-            ]
-        )->shouldBeCalled();
+        $definition->setArguments(Argument::that(function ($params) {
+            $result =
+                $params[0][0] instanceof Reference &&
```
Comment (juliensnz) [inline spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterSerializerPassSpec.php]: I don't understand that

File: spec/Pim/Bundle/DashboardBundle/DependencyInjection/Compiler/RegisterWidgetsPassSpec.php [hunk 3]
```diff
@@ -36,11 +26,23 @@ function it_add_tagged_widgets_to_the_widget_registry(
...
         );
 
-        $factory->createReference('pim_dashboard.widget.foo')->willReturn($fooReference);
-        $factory->createReference('pim_dashboard.widget.bar')->willReturn($barReference);
-
-        $definition->addMethodCall('add', [$fooReference, 10])->shouldBeCalled();
-        $definition->addMethodCall('add', [$barReference, 0])->shouldBeCalled();
+        $definition->addMethodCall('add', Argument::that(function ($params) {
+            return
+                $params[0] instanceof Reference &&
```
Comment (juliensnz) [inline spec/Pim/Bundle/DashboardBundle/DependencyInjection/Compiler/RegisterWidgetsPassSpec.php]: same here, seems weird

File: spec/Pim/Component/Connector/Denormalizer/Flat/ProductValue/AttributeOptionsDenormalizerSpec.php [hunk 4]
```diff
@@ -24,14 +24,26 @@ function it_is_a_denormalizer()
...
-        $denormalizer->denormalize('2', 'pim_catalog_simpleselect', Argument::cetera())->shouldBeCalled()->willReturn($blue);
-        $denormalizer->denormalize('3', 'pim_catalog_simpleselect', Argument::cetera())->shouldBeCalled()->willReturn($green);
+        $denormalizer->denormalize('1', 'pim_catalog_simpleselect', Argument::cetera())
+            ->shouldBeCalled()
+            ->willReturn($red);
+        $denormalizer->denormalize('2', 'pim_catalog_simpleselect', Argument::cetera())
+            ->shouldBeCalled()
+            ->willReturn($blue);
+        $denormalizer->denormalize('3', 'pim_catalog_simpleselect', Argument::cetera())
+            ->shouldBeCalled()
```
Comment (juliensnz) [inline spec/Pim/Component/Connector/Denormalizer/Flat/ProductValue/AttributeOptionsDenormalizerSpec.php]: please remove that

File: spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterSerializerPassSpec.php (hunk 2 above)
Comment (aRn0D) [inline spec/Pim/Bundle/CatalogBundle/DependencyInjection/Compiler/RegisterSerializerPassSpec.php]: We remove the `ReferenceFactory` which are created to spec easy not for a real need. That is way to spec a compiler pass which create `Reference` instance.

Here `$params[0]` is `[$barNormalizer, $fooNormalizer]` and `$params[0][0]` is `$barNormalizer`. Thus, I check that is `$barNormalizer` a `Reference` and its identifier is 'normalizer.foo'

File: spec/Pim/Bundle/DashboardBundle/DependencyInjection/Compiler/RegisterWidgetsPassSpec.php (hunk 3 above)
Comment (aRn0D) [inline spec/Pim/Bundle/DashboardBundle/DependencyInjection/Compiler/RegisterWidgetsPassSpec.php]: `$params[0]` =>fooReference (reference)
`$params[1]` => 10 (priority)

File: spec/Pim/Component/Connector/Denormalizer/Flat/ProductValue/AttributeOptionsDenormalizerSpec.php (hunk 4 above)
Comment (aRn0D) [inline spec/Pim/Component/Connector/Denormalizer/Flat/ProductValue/AttributeOptionsDenormalizerSpec.php]: These lines exceed more than 120 characters.

**Comment by SamirBoulil** [comment]
Wow dude !! Good work ! :+1:  nice :fire: !

I still have those questions: 
- `Pim\Component\Connector\Encoder\CsvEncoderSpec.php`: you could rename the method starting with "it_encode_\*" by "it_encode**s**_*"
- `Pim\Component\Connector\ArrayConverter\Structured\JobInstanceConverter.php`:28: what is the difference between the before structure and the after structure ? What is relevant ?
- `Pim\Bundle\BaseConnectorBundle\Reader\Doctrine\ODMProductReader.php`: 68 align comments ? Should we use FQCN in the attribute comment ?

Good job ;)

**Comment by aRn0D** [comment]
@SamirBoulil every converters check and convert data, this one only check and does not convert them. I use this way to be consistent with other converters.

File: app/AppKernel.php [hunk 5]
```diff
@@ -84,7 +84,6 @@ protected function getPimBundles()
             new Pim\Bundle\WebServiceBundle\PimWebServiceBundle(),
             new Pim\Bundle\EnrichBundle\PimEnrichBundle(),
             new Pim\Bundle\BaseConnectorBundle\PimBaseConnectorBundle(),
-            new Pim\Bundle\TransformBundle\PimTransformBundle(),
```
Comment (pierallard) [inline app/AppKernel.php]: :dancer: oh yeah

**Pull request closed without merging**